### PR TITLE
Added new integration test to commit offset manually and an small refactor

### PR DESCRIPTION
**Pull request by franvila**
### Type of change

- Enhancement / new feature

### Description

Added new integration test to commit offset manually and an small refactor

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -93,7 +93,7 @@
             <timezone>Pacific/Auckland</timezone>
         </developer>
         <developer>
-            <name>Francisco Vila</name>
+            <name>Fran Vila</name>
```
Comment (SamBarker) [inline pom.xml:96]: Sorry!

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java [hunk 2]
```diff
@@ -203,9 +203,65 @@ public void shouldModifyProduceMessage(KafkaCluster cluster, Admin admin) throws
...
+            long actualEndOffset;
+            try (var consumer = new KafkaConsumer<String, String>(Map.of(
+                    ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, PROXY_ADDRESS,
+                    ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class,
+                    ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class,
+                    ConsumerConfig.GROUP_ID_CONFIG, "my-group-id",
+                    ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false",
+                    ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest"))) {
+
+                Set<TopicPartition> partitions = new HashSet<>();
```
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:244]: I think this is redundant.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java [hunk 3]
```diff
@@ -203,9 +203,65 @@ public void shouldModifyProduceMessage(KafkaCluster cluster, Admin admin) throws
...
+
+                consumer.commitSync();
+
+                var committed = consumer.committed(partitions);
+                OffsetAndMetadata offsetAndMetadata = committed.get(actualTopicPartition);
+                offsetCommitted = offsetAndMetadata.offset();
+                actualEndOffset = consumer.endOffsets(partitions).get(actualTopicPartition);
+            }
+            assertEquals(offsetCommitted, actualEndOffset);
+            assertEquals(PLAINTEXT, records1.iterator().next().value());
```
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:259]: I wonder is there a way to validate the offsets against what's in the physical cluster as well. I really want to validate that the proxy is handling them correctly. So not just validating that it is applying a consistent modification but that its passing on something valid as well.

File: .gitignore [hunk 4]
```diff
@@ -16,78 +16,79 @@ bin/
 # Reference: https://intellij-support.jetbrains.com/hc/en-us/articles/206544839
 # Copied from https://github.com/github/gitignore/blob/4488915eec0b3a45b5c63ead28f286819c0917de/Global/JetBrains.gitignore
 
- # User-specific stuff
```
Comment (SamBarker) [inline .gitignore:19]: Any idea what changed here?

**Review by SamBarker — COMMENTED** [review]
Looking good. 

Would like to see some discussion about validating the offset handling on the underlying cluster before approving (we don't necessarily need to fix that in this PR)

File: .gitignore (hunk 4 above)
Comment (franvila) [reply .gitignore:19]: Removed one white space at the beginning for all the lines

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java [hunk 5]
```diff
@@ -121,20 +123,20 @@ public void shouldPassThroughRecordUnchanged(KafkaCluster cluster, Admin admin)
...
         try (var proxy = startProxy(config)) {
             try (var producer = new KafkaProducer<String, String>(Map.of(
-                    ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, proxyAddress,
+                    ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, PROXY_ADDRESS,
                     ProducerConfig.CLIENT_ID_CONFIG, "shouldPassThroughRecordUnchanged",
                     ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class,
                     ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class,
                     ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG, 3_600_000))) {
-                producer.send(new ProducerRecord<>(TOPIC_1, "my-key", "Hello, world!")).get();
+                producer.send(new ProducerRecord<>(TOPIC_1, "my-key", PLAINTEXT)).get();
```
[the remark this answers is not in the document]
Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:135]: Totally agree. Changed

File: pom.xml (hunk 1 above)
Comment (franvila) [reply pom.xml:96]: No worries

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java (hunk 2 above)
Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:244]: What is redundant? Setting it to `earliest` or setting any value into `AUTO_OFFSET_RESET_CONFIG`?

If we don't set a value there, the consumer throws an exception because it needs to have a value there. [see here](https://medium.com/lydtech-consulting/kafka-consumer-auto-offset-reset-d3962bad2665)

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java [hunk 6]
```diff
@@ -203,9 +203,65 @@ public void shouldModifyProduceMessage(KafkaCluster cluster, Admin admin) throws
...
+                clusters:
+                  demo:
+                    bootstrap_servers: %s
+                filters:
+                - type: ApiVersions
+                - type: BrokerAddress
+                """.formatted(PROXY_ADDRESS, cluster.getBootstrapServers());
+
+        try (var proxy = startProxy(config)) {
+            try (var producer = new KafkaProducer<String, String>(Map.of(
```
[the remark this answers is not in the document]
Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:222]: Here for the producer we need to set the bootstrap server using the `kroxy`'s, but by default it shall use the cluster's one. I don't know if there is any annotation to do that, but I don't find the way to say to the producer of the extension to modify the default bootstrap server.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java [hunk 7]
```diff
@@ -203,9 +203,65 @@ public void shouldModifyProduceMessage(KafkaCluster cluster, Admin admin) throws
...
+                    ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG, 3_600_000))) {
+                producer.send(new ProducerRecord<>(TOPIC_1, "my-key", PLAINTEXT)).get();
+                producer.send(new ProducerRecord<>(TOPIC_1, "my-key2", PLAINTEXT)).get();
+                producer.flush();
+            }
+
+            ConsumerRecords<String, String> records1;
+            long offsetCommitted;
+            long actualEndOffset;
+            try (var consumer = new KafkaConsumer<String, String>(Map.of(
```
[the remark this answers is not in the document]
Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:236]: Here for the consumer we need to set the bootstrap server using the `kroxy`'s, but by default it shall use the cluster's one. I don't know if there is any annotation to do that, but I don't find the way to say to the consumer of the extension to modify the default bootstrap server.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:222]: Ahh. Ok should have twigged that. I think we should workout how to do that better but that's a story for another day.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:236]: Yeah. As above your right, we do need a better way of handling Proxies and their clients in the tests

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java (hunk 3 above)
Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java]: @showuon could you help us with this?

Comment (showuon) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java]: Will take a look today or tomorrow.

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java [hunk 8]
```diff
@@ -203,9 +203,65 @@ public void shouldModifyProduceMessage(KafkaCluster cluster, Admin admin) throws
         }
     }
 
+    @Test
+    public void offsetManuallyManaged(KafkaCluster cluster, Admin admin) throws Exception {
```
Comment (k-wall) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: I don't understand the motivation for adding this test.  The ApiVersion/BrokerAddress filters don't care about offsets in any way, so why is this test important to have.

Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: I asked on the slack channel and it was suggested by @SamBarker, having some test for a manually managed offset.

Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: I am looking for a test that validates that kroxylicious is not messing with offset handling (i.e. normal operation everything works) as a safety net for future changes where it does manipulate the requests. 

I think there are several scenarios that need covered.
1. General usage is unaffected clients can manage offsets normally when kroxylicious *is not manipulating* the commit
2. A test that verifies what gets written to the physical cluster is identical to what was committed to the virtual cluster
3. A test that verifies what gets written to the physical cluster is different but equivalent to what was committed to the virtual cluster.
4. General usage is unaffected clients can manage offsets normally when kroxylicious *is manipulating* the commit 

Which when I enumerate them like that is really two themes. Tests validate Clients of the virtual cluster operate normally regardless of what the proxy is doing. Tests validate that Kroxylicious is transforming the message as intended by the filter set.

My take is that should be our standard "template" for each API or Flow we want to support (I'm not sure what level of abstraction makes sense to focus those around).

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java [hunk 9]
```diff
@@ -203,9 +203,65 @@ public void shouldModifyProduceMessage(KafkaCluster cluster, Admin admin) throws
...
+                records1 = consumer.poll(Duration.ofSeconds(100));
+
+                consumer.commitSync();
+
+                var committed = consumer.committed(partitions);
+                OffsetAndMetadata offsetAndMetadata = committed.get(actualTopicPartition);
+                offsetCommitted = offsetAndMetadata.offset();
+                actualEndOffset = consumer.endOffsets(partitions).get(actualTopicPartition);
+            }
+            assertEquals(offsetCommitted, actualEndOffset);
```
Comment (showuon) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:258]: This is not correct, actually. You are verifying kafka is storing committed offset correctly via `committed` and `endOffsets` API, but not verifying we're committing the expected offsets. We can use this [API](https://javadoc.io/static/org.apache.kafka/kafka-clients/3.4.0/org/apache/kafka/clients/consumer/KafkaConsumer.html#commitSync-java.util.Map-) to commit our "expected offset". And then retrieve the "actual offset" from broker via `committed` API, to make sure kroxy is working as expected.

Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java]: Thanks @showuon! Changed using the suggested API. Please, could you review it again and see if I misunderstood anything?

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java [hunk 10]
```diff
@@ -203,9 +203,67 @@ public void shouldModifyProduceMessage(KafkaCluster cluster, Admin admin) throws
...
+
+                actualEndOffset = consumer.endOffsets(partitions).get(actualTopicPartition);
+                offsetToCommit = actualEndOffset - 1;
+                consumer.commitSync(Map.of(actualTopicPartition, new OffsetAndMetadata(offsetToCommit)));
+
+                var committed = consumer.committed(partitions);
+                OffsetAndMetadata offsetAndMetadata = committed.get(actualTopicPartition);
+                offsetCommitted = offsetAndMetadata.offset();
+            }
+            assertEquals(offsetCommitted, offsetToCommit);
```
Comment (showuon) [inline integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:260]: Yes, this is pretty much what I expected. I'd refactor it a little bit.
1. We don't need to poll for records, actually. All we need is to commit an offset to broker
2. Verify the offsetCommitted is changed after we committed (i.e., make sure kproxy indeed pass the correct info to the broker)

```java
                // before committing the offset, make sure the offset stored in broker is 0 (or -1 or exception (maybe assertThrow to verify?))
                var beforeCommitted = consumer.committed(partitions);
                assertEquals(0, beforeCommitted);

                // hard code the commit to offset
                offsetToCommit = 2;
                consumer.commitSync(Map.of(actualTopicPartition, new OffsetAndMetadata(offsetToCommit)));

                var committed = consumer.committed(partitions);
                OffsetAndMetadata offsetAndMetadata = committed.get(actualTopicPartition);
                offsetCommitted = offsetAndMetadata.offset();
            }
            assertEquals(offsetCommitted, offsetToCommit);
```

WDYT?

Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:260]: Great idea to assert before commit the offset. Thanks!

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java (hunk 8 above)
Comment (k-wall) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: > I am looking for a test that validates that kroxylicious is not messing with offset handling 

what makes you pick out offset handling for requiring extra tests?  why not say, aspects of the consumer group protocol?  If we have a offsetmaniluatingfilter then sure we'd want test cases around that but as there is nothing in kroxyilicious's core manipulates offsets, I'm failing to get your motivation.

Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: I'm looking for it any and everywhere 😁  I don't intend to be specific to offset handling, I was just wanting to establish the pattern here.

**Review by SamBarker — APPROVED** [review]
LGTM

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java (hunk 8 above)
Comment (k-wall) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: I'm sorry. I still don't get it.  The filter chain under test is ApiVersions/BrokerAddress.  Kroxy isn't even going to decode the RPCs related to offsets. Kroxy is nothing more than a boring old byte pipe.  I don't understand the value that this test brings.

**Comment by k-wall** [comment]
I'm voting -1 on this for the reasons I express above.

**Comment by franvila** [comment]
So... What should we do?

File: integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java (hunk 8 above)
Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: In the short term validating that it is a boring pipe seems valuable to me. 

I don't understand we we wouldn't want test coverage that the pipe does what we expect, even if that is don't change it. 

IIRC we are/going to re-write clientids and groupIds which must be part of what is committed to Kafka, in multi tenant land. It seems valuable to me that should that handling go wrong we have a test demonstrating that the basics work.

Comment (k-wall) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: The MultiTenantITs are defintely going to cover offsets.  Those tests do that already (to a small extent), and we are in the process of expanding that to cover other offset use-cases.

Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: > Where would we stop? Would we test all right tests for every unfiltered RPC?
Yes. Baseline proof that the proxy doesn't break the individual RPC's. I do not understand how that is remotely controversial. I'm not advocating for in-depth coverage or testing every conceivable edge case just a basic smoke test that each RPC is proxied correctly. 

> The [MultiTenantITs](https://github.com/search?q=repo%3Akroxylicious%2Fkroxylicious%20consumeOneAndOffsetCommit&type=code) are definitely going to cover offsets (and client id and group ids). That's justified because it is intercepting those RPCs. Those tests already cover offsets (to a small extent), and we are in the process of growing that to cover other offset use-cases right now.

So when one of those breaks I think there is value in knowing that the basic pipe doesn't affect things.

Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: > So when one of those breaks I think there is value in knowing that the basic pipe doesn't affect things.

Totally agree with Sam here.

Comment (k-wall) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: My point is kroxy (in KrpcFilterIT configuration) is not decoding the RPCs used by the test cases you are proposing writing. No filters will be  firing. No extra code running.  All you are exercising is the kroxy's ability to shunt opaque bytes to/from the network (there's a correlation id rewrite, but that's non RPC specific common code).  The integration test is already demonstrating that shunting bytes works.   Testing it will n different RPCs brings us zero.

This to me is like saying a washing machine isn't tested until you've washed a sock of every colour.

Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: >Testing it will n different RPCs brings us zero.

Ok that's the fundamental disconnect. I'm saying that it brings us value. It might be small in general but when something breaks in a specific filter validating that the rest of the proxy is not causing issues.

A couple of possible ways to break the impasse:
 1.  Add a NoOp filter which touches every RPC so that we are exercising the decode encode logic
 2. Add a Filter which adds a header to every RPC to mark it as having gone through the proxy so that the proxy under test is actually doing something.
 
 
 >This to me is like saying a washing machine isn't tested until you've washed a sock of every colour.
 
 To use your analogy I think it's akin to saying that the washing machine isn't tested until it has washed a bed sheet, a towel, a t-shirt a sock, and a silk shirt. The wash process is consistent between them but we are validating they all survive the wash cycle.

Comment (k-wall) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: Demonstrating fidelity in the decode encode logic seems like a valuable test.  However, I don't think using the client/broker is the best way to achieve that.   Couldn't we write a test that does that without the kafka client/broker?  Either as a unit test where we just round-trip decode->encode, and/or an network integrated test that just uses a fake client to send all possible RPCs/versions and a fake server that validated what is received (and the reciprocal for responses).  Using code generation for the tests might be worth considering.

Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: With #243 merged we now have this coverage of Kroxylicious decoding/encoding/intercepting/proxying all RPCs at each supported apiVersion. Does this unblock this PR?

Comment (k-wall) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: In my mind, the work on #243 makes the case for closing this PR (without merge) stronger.

Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: @franvila are you happy that #243 covers everything from here or is there still milage in this PR? 

My expectation is that #243 has things covered and we should close this one.

Comment (franvila) [reply integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:207]: Agree. Thanks @robobario!
